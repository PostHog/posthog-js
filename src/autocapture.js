import { _ } from './utils'
import {
    getClassName,
    getSafeText,
    isElementNode,
    isTag,
    isTextNode,
    shouldCaptureDomEvent,
    shouldCaptureElement,
    shouldCaptureValue,
    usefulElements,
} from './autocapture-utils'

var autocapture = {
    _initializedTokens: [],

    _previousElementSibling: function (el) {
        if (el.previousElementSibling) {
            return el.previousElementSibling
        } else {
            do {
                el = el.previousSibling
            } while (el && !isElementNode(el))
            return el
        }
    },

    _loadScript: function (scriptUrlToLoad, callback) {
        var scriptTag = document.createElement('script')
        scriptTag.type = 'text/javascript'
        scriptTag.src = scriptUrlToLoad
        scriptTag.onload = callback

        var scripts = document.getElementsByTagName('script')
        if (scripts.length > 0) {
            scripts[0].parentNode.insertBefore(scriptTag, scripts[0])
        } else {
            document.body.appendChild(scriptTag)
        }
    },

    _getPropertiesFromElement: function (elem) {
        var tag_name = elem.tagName.toLowerCase()
        var props = {
            tag_name: tag_name,
        }
        if (usefulElements.indexOf(tag_name) > -1) props['$el_text'] = getSafeText(elem)

        var classes = getClassName(elem)
        if (classes.length > 0) props['classes'] = classes.split(' ')

        if (shouldCaptureElement(elem)) {
            _.each(elem.attributes, function (attr) {
                if (shouldCaptureValue(attr.value)) {
                    props['attr__' + attr.name] = attr.value
                }
            })
        }

        var nthChild = 1
        var nthOfType = 1
        var currentElem = elem
        while ((currentElem = this._previousElementSibling(currentElem))) {
            // eslint-disable-line no-cond-assign
            nthChild++
            if (currentElem.tagName === elem.tagName) {
                nthOfType++
            }
        }
        props['nth_child'] = nthChild
        props['nth_of_type'] = nthOfType

        return props
    },

    _getDefaultProperties: function (eventType) {
        return {
            $event_type: eventType,
            $ce_version: 1,
        }
    },

    _extractCustomPropertyValue: function (customProperty) {
        var propValues = []
        _.each(document.querySelectorAll(customProperty['css_selector']), function (matchedElem) {
            var value

            if (['input', 'select'].indexOf(matchedElem.tagName.toLowerCase()) > -1) {
                value = matchedElem['value']
            } else if (matchedElem['textContent']) {
                value = matchedElem['textContent']
            }

            if (shouldCaptureValue(value)) {
                propValues.push(value)
            }
        })
        return propValues.join(', ')
    },

    _getCustomProperties: function (targetElementList) {
        var props = {}
        _.each(
            this._customProperties,
            function (customProperty) {
                _.each(
                    customProperty['event_selectors'],
                    function (eventSelector) {
                        var eventElements = document.querySelectorAll(eventSelector)
                        _.each(
                            eventElements,
                            function (eventElement) {
                                if (_.includes(targetElementList, eventElement) && shouldCaptureElement(eventElement)) {
                                    props[customProperty['name']] = this._extractCustomPropertyValue(customProperty)
                                }
                            },
                            this
                        )
                    },
                    this
                )
            },
            this
        )
        return props
    },

    _getEventTarget: function (e) {
        // https://developer.mozilla.org/en-US/docs/Web/API/Event/target#Compatibility_notes
        if (typeof e.target === 'undefined') {
            return e.srcElement
        } else {
            return e.target
        }
    },

    _captureEvent: function (e, instance) {
        /*** Don't mess with this code without running IE8 tests on it ***/
        var target = this._getEventTarget(e)
        if (isTextNode(target)) {
            // defeat Safari bug (see: http://www.quirksmode.org/js/events_properties.html)
            target = target.parentNode
        }

        if (shouldCaptureDomEvent(target, e)) {
            var targetElementList = [target]
            var curEl = target
            while (curEl.parentNode && !isTag(curEl, 'body')) {
                targetElementList.push(curEl.parentNode)
                curEl = curEl.parentNode
            }

            var elementsJson = []
            var href,
                explicitNoCapture = false
            _.each(
                targetElementList,
                function (el) {
                    var shouldCaptureEl = shouldCaptureElement(el)

                    // if the element or a parent element is an anchor tag
                    // include the href as a property
                    if (el.tagName.toLowerCase() === 'a') {
                        href = el.getAttribute('href')
                        href = shouldCaptureEl && shouldCaptureValue(href) && href
                    }

                    // allow users to programmatically prevent capturing of elements by adding class 'ph-no-capture'
                    var classes = getClassName(el).split(' ')
                    if (_.includes(classes, 'ph-no-capture')) {
                        explicitNoCapture = true
                    }

                    elementsJson.push(this._getPropertiesFromElement(el))
                },
                this
            )

            elementsJson[0]['$el_text'] = getSafeText(target)

            if (href) {
                elementsJson[0]['attr__href'] = href
            }

            if (explicitNoCapture) {
                return false
            }

            // only populate text content from target element (not parents)
            // to prevent text within a sensitive element from being collected
            // as part of a parent's el.textContent
            var elementText
            var safeElementText = getSafeText(target)
            if (safeElementText && safeElementText.length) {
                elementText = safeElementText
            }

            var props = _.extend(
                this._getDefaultProperties(e.type),
                {
                    $elements: elementsJson,
                },
                this._getCustomProperties(targetElementList)
            )

            instance.capture('$autocapture', props)
            return true
        }
    },

    // only reason is to stub for unit tests
    // since you can't override window.location props
    _navigate: function (href) {
        window.location.href = href
    },

    _addDomEventHandlers: function (instance) {
        var handler = _.bind(function (e) {
            e = e || window.event
            this._captureEvent(e, instance)
        }, this)
        _.register_event(document, 'submit', handler, false, true)
        _.register_event(document, 'change', handler, false, true)
        _.register_event(document, 'click', handler, false, true)
    },

    _customProperties: {},
    init: function (instance) {
        if (this._maybeLoadEditor(instance)) return // don't autocapture actions when the editor is enabled

        var token = instance.get_config('token')
        if (this._initializedTokens.indexOf(token) > -1) {
            console.log('autocapture already initialized for token "' + token + '"')
            return
        }
        this._initializedTokens.push(token)

        var parseDecideResponse = _.bind(function (response) {
            if (!(document && document.body)) {
                console.log('document not ready yet, trying again in 500 milliseconds...')
                setTimeout(function () {
                    parseDecideResponse(response)
                }, 500)
                return
            }
            var editorParams =
                response['editorParams'] ||
                (response['toolbarVersion'] ? { toolbarVersion: response['toolbarVersion'] } : {})
            if (
                response['isAuthenticated'] &&
                editorParams['toolbarVersion'] &&
                editorParams['toolbarVersion'].indexOf('toolbar') === 0
            ) {
                this._loadEditor(
                    instance,
                    Object.assign({}, editorParams, {
                        apiURL: instance.get_config('api_host'),
                    })
                )
                instance.set_config({ debug: true })
            }
            if (response && response['config'] && response['config']['enable_collect_everything'] === true) {
                if (response['custom_properties']) {
                    this._customProperties = response['custom_properties']
                }
                this._addDomEventHandlers(instance)
            } else {
                instance['__autocapture_enabled'] = false
            }

            if (response['featureFlags'] && response['featureFlags'].length > 0) {
                instance.persistence.register({ $active_feature_flags: response['featureFlags'] })
            } else {
                instance.persistence.unregister('$active_feature_flags')
            }
        }, this)

        var json_data = _.JSONEncode({
            token: token,
            distinct_id: instance.get_distinct_id(),
        })
        var encoded_data = _.base64Encode(json_data)
        instance._send_request(
            instance.get_config('api_host') + '/decide/',
            { data: encoded_data },
            { method: 'POST' },
            instance._prepare_callback(parseDecideResponse)
        )
    },

    _editorParamsFromHash: function (instance, hash) {
        try {
            var state = _.getHashParam(hash, 'state')
            return JSON.parse(decodeURIComponent(state))
        } catch (e) {
            console.error('Unable to parse data from hash', e)
        }
        return {}
    },

    /**
     * To load the visual editor, we need an access token and other state. That state comes from one of three places:
     * 1. In the URL hash params if the customer is using an old snippet
     * 2. From session storage under the key `editorParams` if the editor was initialized on a previous page
     */
    _maybeLoadEditor: function (instance) {
        try {
            var stateHash = _.getHashParam(window.location.hash, 'state')
            var state = stateHash ? JSON.parse(decodeURIComponent(stateHash)) : null
            var parseFromUrl = state && state['action'] === 'mpeditor'
            var editorParams

            if (parseFromUrl) {
                // happens if they are initializing the editor using an old snippet
                editorParams = state

                if (editorParams && Object.keys(editorParams).length > 0) {
                    window.sessionStorage.setItem('_postHogEditorParams', JSON.stringify(editorParams))

                    if (state['desiredHash']) {
                        // hash that was in the url before the redirect
                        window.location.hash = state['desiredHash']
                    } else if (window.history) {
                        history.replaceState('', document.title, window.location.pathname + window.location.search) // completely remove hash
                    } else {
                        window.location.hash = '' // clear hash (but leaves # unfortunately)
                    }
                }
            } else {
                // get credentials from sessionStorage from a previous initialzation
                editorParams = JSON.parse(window.sessionStorage.getItem('_postHogEditorParams') || '{}')

                // delete "add-action" or other intent from editorParams, otherwise we'll have the same intent
                // every time we open the page (e.g. you just visiting your own site an hour later)
                delete editorParams.userIntent
            }

            if (editorParams['token'] && instance.get_config('token') === editorParams['token']) {
                this._loadEditor(instance, editorParams)
                return true
            } else {
                return false
            }
        } catch (e) {
            return false
        }
    },

    _loadEditor: function (instance, editorParams) {
        var _this = this
        if (!window['_postHogToolbarLoaded']) {
            // only load the codeless event editor once, even if there are multiple instances of PostHogLib
            window['_postHogToolbarLoaded'] = true
            var host = editorParams['jsURL'] || editorParams['apiURL'] || instance.get_config('api_host')
            var toolbarScript =
                editorParams['toolbarVersion'] && editorParams['toolbarVersion'].indexOf('toolbar') === 0
                    ? 'toolbar.js'
                    : 'editor.js'
            var editorUrl =
                host + (host.endsWith('/') ? '' : '/') + 'static/' + toolbarScript + '?_ts=' + new Date().getTime()
            this._loadScript(editorUrl, function () {
                window['ph_load_editor'](editorParams)
            })
            // Turbolinks doesn't fire an onload event but does replace the entire page, including the toolbar
            _.register_event(window, 'turbolinks:load', function () {
                window['_postHogToolbarLoaded'] = false
                _this._loadEditor(instance, editorParams)
            })
            return true
        }
        return false
    },

    // this is a mechanism to ramp up CE with no server-side interaction.
    // when CE is active, every page load results in a decide request. we
    // need to gently ramp this up so we don't overload decide. this decides
    // deterministically if CE is enabled for this project by modding the char
    // value of the project token.
    enabledForProject: function (token, numBuckets, numEnabledBuckets) {
        numBuckets = !_.isUndefined(numBuckets) ? numBuckets : 10
        numEnabledBuckets = !_.isUndefined(numEnabledBuckets) ? numEnabledBuckets : 10
        var charCodeSum = 0
        for (var i = 0; i < token.length; i++) {
            charCodeSum += token.charCodeAt(i)
        }
        return charCodeSum % numBuckets < numEnabledBuckets
    },

    isBrowserSupported: function () {
        return _.isFunction(document.querySelectorAll)
    },
}

_.bind_instance_methods(autocapture)
_.safewrap_instance_methods(autocapture)

export { autocapture }
