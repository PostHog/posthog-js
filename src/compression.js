import { LZString } from './lz-string'
import { strToU8, gzipSync } from 'fflate'
import { _ } from './utils'

export function decideCompression(compressionSupport) {
    if (compressionSupport['gzip-js']) {
        return 'gzip-js'
    } else if (compressionSupport['lz64']) {
        return 'lz64'
    } else {
        return 'base64'
    }
}

export function compressData(compression, jsonData, options) {
    if (compression === 'lz64') {
        return [{ data: LZString.compressToBase64(jsonData), compression: 'lz64' }, options]
    } else if (compression === 'gzip-js') {
        // :TRICKY: This returns an UInt8Array. We don't encode this to a string - returning a blob will do this for us.
        return [
            gzipSync(strToU8(jsonData), { mtime: 0 }),
            { ...options, blob: true, urlQueryArgs: { compression: 'gzip-js' } },
        ]
    } else {
        return [{ data: _.base64Encode(jsonData) }, options]
    }
}
