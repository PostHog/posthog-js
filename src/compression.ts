import { gzipSync, strToU8 } from 'fflate'
import { _base64Encode } from './utils'
import { Compression, CompressionData, XHROptions } from './types'

export function decideCompression(compressionSupport: Partial<Record<Compression, boolean>>): Compression {
    if (compressionSupport[Compression.GZipJS]) {
        return Compression.GZipJS
    } else {
        return Compression.Base64
    }
}

export function compressData(
    compression: Compression,
    jsonData: string,
    options: XHROptions
): [CompressionData | Uint8Array, XHROptions] {
    if (compression === Compression.GZipJS) {
        // :TRICKY: This returns an UInt8Array. We don't encode this to a string - returning a blob will do this for us.
        return [
            gzipSync(strToU8(jsonData), { mtime: 0 }),
            { ...options, blob: true, urlQueryArgs: { compression: Compression.GZipJS } },
        ]
    } else {
        return [{ data: _base64Encode(jsonData) }, options]
    }
}
