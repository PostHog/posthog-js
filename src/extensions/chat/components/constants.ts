export const BUSINESS_NAME = 'BizPlanner AI'
