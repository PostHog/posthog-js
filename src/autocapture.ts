import {
    _bind_instance_methods,
    _each,
    _extend,
    _includes,
    _isFunction,
    _isUndefined,
    _register_event,
    _safewrap_instance_methods,
    logger,
} from './utils'
import {
    getClassName,
    getSafeText,
    isElementNode,
    isSensitiveElement,
    isTag,
    isTextNode,
    shouldCaptureDomEvent,
    shouldCaptureElement,
    shouldCaptureValue,
    autocaptureCompatibleElements,
    isAngularStyleAttr,
    isDocumentFragment,
    getDirectAndNestedSpanText,
} from './autocapture-utils'
import RageClick from './extensions/rageclick'
import { AutocaptureConfig, AutoCaptureCustomProperty, DecideResponse, Properties } from './types'
import { PostHog } from './posthog-core'
import { AUTOCAPTURE_DISABLED_SERVER_SIDE } from './constants'

const autocapture = {
    _initializedTokens: [] as string[],
    _isDisabledServerSide: null as boolean | null,
    _isAutocaptureEnabled: false as boolean,

    _setIsAutocaptureEnabled: function (instance: PostHog): void {
        const disabled_server_side =
            this._isDisabledServerSide === null
                ? !!instance.persistence?.props[AUTOCAPTURE_DISABLED_SERVER_SIDE]
                : this._isDisabledServerSide
        const enabled_client_side = !!instance.get_config('autocapture')
        this._isAutocaptureEnabled = enabled_client_side && !disabled_server_side
    },

    _previousElementSibling: function (el: Element): Element | null {
        if (el.previousElementSibling) {
            return el.previousElementSibling
        } else {
            let _el: Element | null = el
            do {
                _el = _el.previousSibling as Element | null // resolves to ChildNode->Node, which is Element's parent class
            } while (_el && !isElementNode(_el))
            return _el
        }
    },

    _getAugmentPropertiesFromElement: function (elem: Element): Properties {
        const shouldCaptureEl = shouldCaptureElement(elem)
        if (!shouldCaptureEl) {
            return {}
        }

        const props: Properties = {}

        _each(elem.attributes, function (attr: Attr) {
            if (attr.name.indexOf('data-ph-capture-attribute') === 0) {
                const propertyKey = attr.name.replace('data-ph-capture-attribute-', '')
                const propertyValue = attr.value
                if (propertyKey && propertyValue && shouldCaptureValue(propertyValue)) {
                    props[propertyKey] = propertyValue
                }
            }
        })
        return props
    },

    _getPropertiesFromElement: function (elem: Element, maskInputs: boolean, maskText: boolean): Properties {
        const tag_name = elem.tagName.toLowerCase()
        const props: Properties = {
            tag_name: tag_name,
        }
        if (autocaptureCompatibleElements.indexOf(tag_name) > -1 && !maskText) {
            if (tag_name.toLowerCase() === 'a' || tag_name.toLowerCase() === 'button') {
                props['$el_text'] = getDirectAndNestedSpanText(elem)
            } else {
                props['$el_text'] = getSafeText(elem)
            }
        }

        const classes = getClassName(elem)
        if (classes.length > 0)
            props['classes'] = classes.split(' ').filter(function (c) {
                return c !== ''
            })

        _each(elem.attributes, function (attr: Attr) {
            // Only capture attributes we know are safe
            if (isSensitiveElement(elem) && ['name', 'id', 'class'].indexOf(attr.name) === -1) return

            if (!maskInputs && shouldCaptureValue(attr.value) && !isAngularStyleAttr(attr.name)) {
                props['attr__' + attr.name] = attr.value
            }
        })

        let nthChild = 1
        let nthOfType = 1
        let currentElem: Element | null = elem
        while ((currentElem = this._previousElementSibling(currentElem))) {
            // eslint-disable-line no-cond-assign
            nthChild++
            if (currentElem.tagName === elem.tagName) {
                nthOfType++
            }
        }
        props['nth_child'] = nthChild
        props['nth_of_type'] = nthOfType

        return props
    },

    _getDefaultProperties: function (eventType: string): Properties {
        return {
            $event_type: eventType,
            $ce_version: 1,
        }
    },

    _extractCustomPropertyValue: function (customProperty: AutoCaptureCustomProperty): string {
        const propValues: string[] = []
        _each(document.querySelectorAll(customProperty['css_selector']), function (matchedElem) {
            let value

            if (['input', 'select'].indexOf(matchedElem.tagName.toLowerCase()) > -1) {
                value = matchedElem['value']
            } else if (matchedElem['textContent']) {
                value = matchedElem['textContent']
            }

            if (shouldCaptureValue(value)) {
                propValues.push(value)
            }
        })
        return propValues.join(', ')
    },

    // TODO: delete custom_properties after changeless typescript refactor
    _getCustomProperties: function (targetElementList: Element[]): Properties {
        const props: Properties = {} // will be deleted
        _each(this._customProperties, (customProperty) => {
            _each(customProperty['event_selectors'], (eventSelector) => {
                const eventElements = document.querySelectorAll(eventSelector)
                _each(eventElements, (eventElement) => {
                    if (_includes(targetElementList, eventElement) && shouldCaptureElement(eventElement)) {
                        props[customProperty['name']] = this._extractCustomPropertyValue(customProperty)
                    }
                })
            })
        })
        return props
    },

    _getEventTarget: function (e: Event): Element | null {
        // https://developer.mozilla.org/en-US/docs/Web/API/Event/target#Compatibility_notes
        if (typeof e.target === 'undefined') {
            return (e.srcElement as Element) || null
        } else {
            if ((e.target as HTMLElement)?.shadowRoot) {
                return (e.composedPath()[0] as Element) || null
            }
            return (e.target as Element) || null
        }
    },

    _captureEvent: function (e: Event, instance: PostHog, eventName = '$autocapture'): boolean | void {
        /*** Don't mess with this code without running IE8 tests on it ***/
        let target = this._getEventTarget(e)
        if (isTextNode(target)) {
            // defeat Safari bug (see: http://www.quirksmode.org/js/events_properties.html)
            target = (target.parentNode || null) as Element | null
        }

        if (eventName === '$autocapture' && e.type === 'click' && e instanceof MouseEvent) {
            if (this.rageclicks?.isRageClick(e.clientX, e.clientY, new Date().getTime())) {
                this._captureEvent(e, instance, '$rageclick')
            }
        }

        if (target && shouldCaptureDomEvent(target, e, this.config)) {
            const targetElementList = [target]
            let curEl = target
            while (curEl.parentNode && !isTag(curEl, 'body')) {
                if (isDocumentFragment(curEl.parentNode)) {
                    targetElementList.push((curEl.parentNode as any).host)
                    curEl = (curEl.parentNode as any).host
                    continue
                }
                targetElementList.push(curEl.parentNode as Element)
                curEl = curEl.parentNode as Element
            }

            const elementsJson: Properties[] = []
            const autocaptureAugmentProperties: Properties = {}
            let href,
                explicitNoCapture = false
            _each(targetElementList, (el) => {
                const shouldCaptureEl = shouldCaptureElement(el)

                // if the element or a parent element is an anchor tag
                // include the href as a property
                if (el.tagName.toLowerCase() === 'a') {
                    href = el.getAttribute('href')
                    href = shouldCaptureEl && shouldCaptureValue(href) && href
                }

                // allow users to programmatically prevent capturing of elements by adding class 'ph-no-capture'
                const classes = getClassName(el).split(' ')
                if (_includes(classes, 'ph-no-capture')) {
                    explicitNoCapture = true
                }

                elementsJson.push(
                    this._getPropertiesFromElement(
                        el,
                        instance.get_config('mask_all_element_attributes'),
                        instance.get_config('mask_all_text')
                    )
                )

                const augmentProperties = this._getAugmentPropertiesFromElement(el)
                _extend(autocaptureAugmentProperties, augmentProperties)
            })

            if (!instance.get_config('mask_all_text')) {
                // if the element is a button or anchor tag get the span text from any
                // children and include it as/with the text property on the parent element
                if (target.tagName.toLowerCase() === 'a' || target.tagName.toLowerCase() === 'button') {
                    elementsJson[0]['$el_text'] = getDirectAndNestedSpanText(target)
                } else {
                    elementsJson[0]['$el_text'] = getSafeText(target)
                }
            }

            if (href) {
                elementsJson[0]['attr__href'] = href
            }

            if (explicitNoCapture) {
                return false
            }

            const props = _extend(
                this._getDefaultProperties(e.type),
                {
                    $elements: elementsJson,
                },
                this._getCustomProperties(targetElementList),
                autocaptureAugmentProperties
            )

            instance.capture(eventName, props)
            return true
        }
    },

    // only reason is to stub for unit tests
    // since you can't override window.location props
    _navigate: function (href: string): void {
        window.location.href = href
    },

    _addDomEventHandlers: function (instance: PostHog): void {
        const handler = (e: Event) => {
            e = e || window.event
            this._captureEvent(e, instance)
        }
        _register_event(document, 'submit', handler, false, true)
        _register_event(document, 'change', handler, false, true)
        _register_event(document, 'click', handler, false, true)
    },

    _customProperties: [] as AutoCaptureCustomProperty[],
    rageclicks: null as RageClick | null,
    config: undefined as AutocaptureConfig | undefined,

    init: function (instance: PostHog): void {
        if (typeof instance.__autocapture !== 'boolean') {
            this.config = instance.__autocapture
        }

        // precompile the regex
        if (this.config?.url_allowlist) {
            this.config.url_allowlist = this.config.url_allowlist.map((url) => new RegExp(url))
        }

        this.rageclicks = new RageClick(instance.get_config('rageclick'))
    },

    afterDecideResponse: function (response: DecideResponse, instance: PostHog): void {
        const token = instance.get_config('token')
        if (this._initializedTokens.indexOf(token) > -1) {
            logger.log('autocapture already initialized for token "' + token + '"')
            return
        }

        if (instance.persistence) {
            instance.persistence.register({
                [AUTOCAPTURE_DISABLED_SERVER_SIDE]: !!response['autocapture_opt_out'],
            })
        }
        // store this in-memory incase persistence is disabled
        this._isDisabledServerSide = !!response['autocapture_opt_out']

        this._setIsAutocaptureEnabled(instance)

        this._initializedTokens.push(token)

        if (
            response &&
            response['config'] &&
            response['config']['enable_collect_everything'] &&
            this._isAutocaptureEnabled
        ) {
            // TODO: delete custom_properties after changeless typescript refactor
            if (response['custom_properties']) {
                this._customProperties = response['custom_properties']
            }
            this._addDomEventHandlers(instance)
        } else {
            instance['__autocapture'] = false
        }
    },

    // this is a mechanism to ramp up CE with no server-side interaction.
    // when CE is active, every page load results in a decide request. we
    // need to gently ramp this up so we don't overload decide. this decides
    // deterministically if CE is enabled for this project by modding the char
    // value of the project token.
    enabledForProject: function (
        token: string | null | undefined,
        numBuckets: number,
        numEnabledBuckets: number
    ): boolean {
        if (!token) {
            return true
        }
        numBuckets = !_isUndefined(numBuckets) ? numBuckets : 10
        numEnabledBuckets = !_isUndefined(numEnabledBuckets) ? numEnabledBuckets : 10
        let charCodeSum = 0
        for (let i = 0; i < token.length; i++) {
            charCodeSum += token.charCodeAt(i)
        }
        return charCodeSum % numBuckets < numEnabledBuckets
    },

    isBrowserSupported: function (): boolean {
        return _isFunction(document.querySelectorAll)
    },
}

_bind_instance_methods(autocapture)
_safewrap_instance_methods(autocapture)

export { autocapture }
