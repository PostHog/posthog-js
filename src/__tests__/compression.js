import sinon from 'sinon'
import { autocapture } from '../autocapture'
import { decideCompression, compressData } from '../compression'
import { Decide } from '../decide'
import { AUTOCAPTURE_DISABLED_SERVER_SIDE } from '../constants'

describe('decideCompression()', () => {
    given('subject', () => decideCompression(given.compressionSupport))
    given('compressionSupport', () => ({}))

    it('returns base64 by default', () => {
        expect(given.subject).toEqual('base64')
    })

    it('returns gzip-js if all compressions supported', () => {
        given('compressionSupport', () => ({
            'gzip-js': true,
            'a different thing that is either deprecated or new': true,
        }))

        expect(given.subject).toEqual('gzip-js')
    })

    it('returns base64 if only unexpected compression is received', () => {
        given('compressionSupport', () => ({ 'the new compression that is not supported yet': true }))

        expect(given.subject).toEqual('base64')
    })
})

describe('compressData()', () => {
    given('subject', () => compressData(given.compression, given.jsonData, given.options))

    given('jsonData', () => JSON.stringify({ large_key: new Array(500).join('abc') }))
    given('options', () => ({ method: 'POST' }))

    it('handles base64', () => {
        given('compression', () => 'base64')

        expect(given.subject).toMatchSnapshot()
    })

    it('handles gzip-js', () => {
        given('compression', () => 'gzip-js')

        expect(given.subject).toMatchSnapshot()
    })
})

describe('Payload Compression', () => {
    afterEach(() => {
        document.getElementsByTagName('html')[0].innerHTML = ''
    })

    describe('compression', () => {
        let lib, sandbox

        beforeEach(() => {
            document.title = 'test page'
            sandbox = sinon.createSandbox()
            autocapture._initializedTokens = []
            lib = {
                debug: true,
                _prepare_callback: sandbox.spy((callback) => callback),
                _send_request: sandbox.spy((url, params, options, callback) => {
                    if (url === 'https://test.com/decide/?v=3') {
                        callback({ config: { enable_collect_everything: true }, supportedCompression: ['gzip-js'] })
                    } else {
                        throw new Error('Should not get here')
                    }
                }),
                get_config: sandbox.spy(function (key) {
                    switch (key) {
                        case 'api_host':
                            return 'https://test.com'
                        case 'token':
                            return 'testtoken'
                    }
                }),
                token: 'testtoken',
                get_distinct_id() {
                    return 'distinctid'
                },
                getGroups: () => ({}),

                toolbar: {
                    maybeLoadToolbar: jest.fn(),
                    afterDecideResponse: jest.fn(),
                },
                sessionRecording: {
                    afterDecideResponse: jest.fn(),
                },
                featureFlags: {
                    receivedFeatureFlags: jest.fn(),
                },
                _hasBootstrappedFeatureFlags: jest.fn(),
                get_property: (property_key) =>
                    property_key === AUTOCAPTURE_DISABLED_SERVER_SIDE
                        ? given.$autocapture_disabled_server_side
                        : undefined,
            }
        })
        given('$autocapture_disabled_server_side', () => false)

        afterEach(() => {
            sandbox.restore()
        })

        it('should save supported compression in instance', () => {
            new Decide(lib).call()
            autocapture.init(lib)
            expect(lib.compression).toEqual({ 'gzip-js': true })
        })
    })
})
