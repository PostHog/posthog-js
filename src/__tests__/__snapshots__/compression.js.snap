// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compressData() handles base64 1`] = `
Array [
  Object {
    "data": "eyJsYXJnZV9rZXkiOiJhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmNhYmMifQ==",
  },
  Object {
    "method": "POST",
  },
]
`;

exports[`compressData() handles gzip-js 1`] = `
Array [
  Uint8Array [
    31,
    139,
    8,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
    171,
    86,
    202,
    73,
    44,
    74,
    79,
    141,
    207,
    78,
    173,
    84,
    178,
    82,
    74,
    76,
    74,
    30,
    69,
    163,
    104,
    20,
    141,
    162,
    81,
    52,
    156,
    144,
    82,
    45,
    0,
    150,
    15,
    165,
    86,
    233,
    5,
    0,
    0,
  ],
  Object {
    "blob": true,
    "method": "POST",
    "urlQueryArgs": Object {
      "compression": "gzip-js",
    },
  },
]
`;
